yarn-project/types/src/l2_block.ts

```diff
@@ -224,6 +224,130 @@ export class L2Block {
+      this.startPublicDataTreeRoot,
+      this.endPrivateDataTreeSnapshot,
+      this.endNullifierTreeSnapshot,
+      this.endContractTreeSnapshot,
+      this.endTreeOfHistoricPrivateDataTreeRootsSnapshot,
+      this.endTreeOfHistoricContractTreeRootsSnapshot,
+      this.endPublicDataTreeRoot,
+      this.getCalldataHash(),
+    );
+    const temp = toBigIntBE(sha256(buf));
+    const p = BigInt('21888242871839275222246405745257275088548364400416034343698204186575808495617');
+    return Fr.fromBuffer(toBufferBE(temp % p, 32));
+  }
+
+  /**
+   * Computes the start state hash (should equal contract data before block)
+   * @returns The start state hash for the L2 block
+   */
+  getStartStateHash() {
+    const inputValue = serializeToBuffer(
+      this.number - 1,
+      this.startPrivateDataTreeSnapshot,
+      this.startNullifierTreeSnapshot,
+      this.startContractTreeSnapshot,
+      this.startTreeOfHistoricPrivateDataTreeRootsSnapshot,
+      this.startTreeOfHistoricContractTreeRootsSnapshot,
+      this.startPublicDataTreeRoot,
+    );
+    return sha256(inputValue);
+  }
+
+  /**
+   * Computes the end state hash (should equal contract data after block)
+   * @returns The end state hash for the L2 block
+   */
+  getEndStateHash() {
+    const inputValue = serializeToBuffer(
+      this.number,
+      this.endPrivateDataTreeSnapshot,
+      this.endNullifierTreeSnapshot,
+      this.endContractTreeSnapshot,
+      this.endTreeOfHistoricPrivateDataTreeRootsSnapshot,
+      this.endTreeOfHistoricContractTreeRootsSnapshot,
+      this.endPublicDataTreeRoot,
+    );
+    return sha256(inputValue);
+  }
+
+  /**
+   * Computes the calldata hash for the L2 block
+   * This calldata hash is also computed by the rollup contract when the block is submitted,
+   * and inside the circuit, it is part of the public inputs.
+   * @returns The calldata hash.
+   */
+  getCalldataHash() {
+    const computeRoot = (leafs: Buffer[]): Buffer => {
+      const layers: Buffer[][] = [leafs];
+      let activeLayer = 0;
+
+      while (layers[activeLayer].length > 1) {
+        const layer: Buffer[] = [];
+        const layerLength = layers[activeLayer].length;
+
+        for (let i = 0; i < layerLength; i += 2) {
+          const left = layers[activeLayer][i];
+          const right = layers[activeLayer][i + 1];
+
+          layer.push(sha256(Buffer.concat([left, right])));
+        }
+
+        layers.push(layer);
+        activeLayer++;
+      }
+
+      return layers[layers.length - 1][0];
+    };
+
+    const leafCount = this.newCommitments.length / (KERNEL_NEW_COMMITMENTS_LENGTH * 2);
+    const leafs: Buffer[] = [];
+
+    for (let i = 0; i < leafCount; i++) {
+      const commitmentPerBase = KERNEL_NEW_COMMITMENTS_LENGTH * 2;
+      const nullifierPerBase = KERNEL_NEW_NULLIFIERS_LENGTH * 2;
+      const publicDataWritesPerBase = STATE_TRANSITIONS_LENGTH * 2; // @note why is this constant named differently?
```

Comment:
> @note why is this constant named differently?
What name did you expect here?

Reply:
If you look at the commitments and nullifier constants you see that there are `KERNEL_` prefix making it clear that this is per kernel, for `STATE_TRANSITIONS_LENTGH` it is not clear if per kernel, per rollup or whatever it is. Also that it is called `STATE_TRANSITIONS` here but public data writes other places, and state transition could also be the general one from `StartStateHash` to `EndStateHash`.

Reply:
Ok, good catch. In that case replacing `STATE_TRANSITIONS` with `KERNEL_PUB_DATA_WRITES` everywhere sounds good to me. It should probably be done in a separate PR but I don't really care if you do it in the same one. If you are too busy with other things I would have time to pick it up once we merge this.

Reply:
I can do it. Have one for L2 -> L1 stuff that right now is part of my other pr, but might be better to just have one with the naming 🤷

Reply:
@benesjan or @PhilWindle, is there both a function limit and a kernel limit, and do they differ? e.g., for new commitments, there are both a function limit and a kernel limit but the value is the same atm, but have you used the value as a function or kenel?